algorithms/kernel/dtrees/forest/regression/oneapi/cl_kernels/df_batch_regression_kernels.cl

```diff
@@ -0,0 +1,994 @@
+        int curFeatureValue       = -1;
+        int curFeatureId          = -1;
+
+        nodeList[nodeId * nNodeProp + 2] = curFeatureId;
+        nodeList[nodeId * nNodeProp + 3] = curFeatureValue;
+        nodeList[nodeId * nNodeProp + 4] = nRows;
+
+        algorithmFPType mrgN        = (algorithmFPType)0;
+        algorithmFPType mrgMean     = (algorithmFPType)0;
+        algorithmFPType mrgSum2Cent = (algorithmFPType)0;
+
+        algorithmFPType bestRN        = (algorithmFPType)0;
+        algorithmFPType bestRMean     = (algorithmFPType)0;
+        algorithmFPType bestRSum2Cent = (algorithmFPType)0;
+
+        algorithmFPType bestLN        = (algorithmFPType)0;
+        algorithmFPType bestLMean     = (algorithmFPType)0;
+        algorithmFPType bestLSum2Cent = (algorithmFPType)0;
+
+        algorithmFPType mrgRN        = (algorithmFPType)0;
+        algorithmFPType mrgRMean     = (algorithmFPType)0;
+        algorithmFPType mrgRSum2Cent = (algorithmFPType)0;
+
+        algorithmFPType mrgLN        = (algorithmFPType)0;
+        algorithmFPType mrgLMean     = (algorithmFPType)0;
+        algorithmFPType mrgLSum2Cent = (algorithmFPType)0;
+
+        int totalBins = 0;
+
+        // totalBins is calculated by each subgroup
+        for (int featIdx = sub_group_local_id; featIdx < nSelectedFeatures; featIdx += sub_group_size)
+        {
+            int featId = selectedFeatures[nodeId * nSelectedFeatures + featIdx];
+            int nBins  = binOffsets[featId + 1] - binOffsets[featId];
+            totalBins += sub_group_reduce_add(nBins);
+        }
+        totalBins = sub_group_broadcast(totalBins, 0);
+
+        int currFtrIdx  = 0;
+        int featId      = selectedFeatures[nodeId * nSelectedFeatures + currFtrIdx];
+        int binId       = 0;
+        int currFtrBins = binOffsets[featId + 1] - binOffsets[featId];
+        int passedBins  = 0;
+
+        for (int i = local_id; i < totalBins; i += local_size)
+        {
+            while (i >= passedBins + currFtrBins)
+            {
+                passedBins += currFtrBins;
+                currFtrIdx++;
+                featId      = selectedFeatures[nodeId * nSelectedFeatures + currFtrIdx];
+                currFtrBins = binOffsets[featId + 1] - binOffsets[featId];
+            }
+            binId = i - passedBins;
+
+            mrgN        = (algorithmFPType)0;
+            mrgMean     = (algorithmFPType)0;
+            mrgSum2Cent = (algorithmFPType)0;
+
+            mrgRN        = (algorithmFPType)0;
+            mrgRMean     = (algorithmFPType)0;
+            mrgRSum2Cent = (algorithmFPType)0;
+
+            mrgLN        = (algorithmFPType)0;
+            mrgLMean     = (algorithmFPType)0;
+            mrgLSum2Cent = (algorithmFPType)0;
+
+            for (int row = 0; row < nRows; row++)
+            {
+                int id  = treeOrder[rowsOffset + row];
+                int bin = data[id * nFeatures + featId];
+
+                if (bin <= binId)
```

Comment:
Is it possible to remove branching by placing a single call but with arguments set by ternary operators?

Reply:
This can be done if you merge `mrgLN` and `mrgRN` into an array `mrgN[2]`, and pass it like this: `&mrgN[(int)(bin > binID)]`

Same for other variables.

Reply:
thanks, will do.

Reply:
applied your variant